lib/elixir_analyzer/exercise_test/common_checks/private_helper_functions.ex

```diff
@@ -0,0 +1,57 @@
+defmodule ElixirAnalyzer.ExerciseTest.CommonChecks.PrivateHelperFunctions do
+  @moduledoc """
+  Compare the module functions to the exemploid and check if helper functions should be private.
+  """
+
+  alias ElixirAnalyzer.Constants
+  alias ElixirAnalyzer.Comment
+
+  @def_ops [:def, :defp, :defmacro, :defmacrop, :defguard, :defguardp]
+  @public_ops [:def, :defmacro, :defguard]
+
+  @spec run(Macro.t(), nil | Macro.t()) :: [{:pass | :fail | :skip, %Comment{}}]
+  def run(_ast, nil), do: []
+
+  def run(code_ast, exemplar_ast) do
+    {_, code_definitions} = Macro.prewalk(code_ast, MapSet.new(), &traverse/2)
+    {_, exemploid_definitions} = Macro.prewalk(exemplar_ast, MapSet.new(), &traverse/2)
+
+    case find_public_helpers(code_definitions, exemploid_definitions) do
+      [] ->
+        []
+
+      [{wrong_definition, correct_definition} | _] ->
+        [
+          {:fail,
+           %Comment{
+             type: :informative,
+             name: Constants.solution_private_helper_functions(),
+             comment: Constants.solution_private_helper_functions(),
+             params: %{
+               expected: correct_definition,
+               actual: wrong_definition
+             }
+           }}
+        ]
+    end
+  end
+
+  defp traverse({op, _meta, [{:when, _, [{name, _, _} | _]} | _]} = ast, names)
+       when op in @def_ops do
+    {ast, MapSet.put(names, {op, name})}
+  end
+
+  defp traverse({op, _meta, [{name, _meta2, _arguments} | _]} = ast, names) when op in @def_ops do
+    {ast, MapSet.put(names, {op, name})}
+  end
```

Comment:
I'm seeing that arity is not used in this check, which makes me wonder what will happen in case there is a "helper" that should have been private but it happens to have the same name as a function that has to be public, only different arity.

```elixir
# solution
def add(a, b), do: add(a, b, 0)
def add(a, b, c), do: a + b + c

# exemplar
def add(a, b), do: a + b
```

Reply:
Good idea, I will distinguish the arities.